src/execution/index/art/leaf.cpp

```diff
@@ -6,45 +6,104 @@
 #include <cstring>
 
 namespace duckdb {
+idx_t Leaf::GetCapacity() const {
+	return IsInlined() ? 1 : rowids.ptr[0];
+}
+
+bool Leaf::IsInlined() const {
+	return count <= 1;
+}
+
+row_t Leaf::GetRowId(idx_t index) {
+	D_ASSERT(index < count);
+	if (IsInlined()) {
+		return rowids.inlined;
+	} else {
+		D_ASSERT(rowids.ptr[0] >= count);
+		return rowids.ptr[index + 1];
+	}
+}
+
+row_t *Leaf::GetRowIds() {
+	if (IsInlined()) {
+		return &rowids.inlined;
+	} else {
+		return rowids.ptr + 1;
+	}
+}
 
 Leaf::Leaf(Key &value, uint32_t depth, row_t row_id) : Node(NodeType::NLeaf) {
-	capacity = 1;
-	row_ids = unique_ptr<row_t[]>(new row_t[capacity]);
-	row_ids[0] = row_id;
 	count = 1;
+	rowids.inlined = row_id;
 	D_ASSERT(value.len >= depth);
 	prefix = Prefix(value, depth, value.len - depth);
 }
 
-Leaf::Leaf(Key &value, uint32_t depth, unique_ptr<row_t[]> row_ids_p, idx_t num_elements_p) : Node(NodeType::NLeaf) {
-	capacity = num_elements_p;
-	row_ids = move(row_ids_p);
+Leaf::Leaf(Key &value, uint32_t depth, row_t *row_ids_p, idx_t num_elements_p) : Node(NodeType::NLeaf) {
+	D_ASSERT(num_elements_p >= 1);
+	if (num_elements_p == 1) {
+		// we can inline the row ids
+		rowids.inlined = row_ids_p[0];
+	} else {
+		// new row ids of this leaf
+		count = 0;
```

Comment:
Why do we set the count to 0 here if we later do `count = num_elements_p`?

Reply:
`Resize` deletes the current array if one exists, so we need to initialize count to 0 to avoid causing an invalid delete